File: zroky-sdk/zroky/intent.py
# SPDX-License-Identifier: FSL-1.1-MIT
# Copyright 2026 Zroky AI

from __future__ import annotations

from typing import Any, Mapping
from uuid import uuid4

import httpx

from zroky._errors import (
    ZrokyRuntimePolicyApprovalRequired,
    ZrokyRuntimePolicyBlocked,
    ZrokyRuntimePolicyError,
)
from zroky._verified_action import _api_url, _ensure_config, _headers

_REQUEST_TIMEOUT_S = 8.0


def pre_execution_guard(
    *,
    intent: Mapping[str, Any],
    environment: str = "production",
    agent_ref: str | None = None,
    idempotency_key: str | None = None,
) -> dict[str, Any]:
    """Create a trusted intent and enforce final policy before side effects."""
    cfg = _ensure_config()
    key = idempotency_key or f"intent-{uuid4()}"
    try:
        intent_response = httpx.post(
            _api_url(cfg.ingest_url, "/v1/intents"),
            headers=_headers(cfg, idempotency_key=key),
            json={"environment": environment, "agent_ref": agent_ref, "intent": dict(intent)},
            timeout=_REQUEST_TIMEOUT_S,
        )
        intent_response.raise_for_status()
        created_intent = intent_response.json()

        policy_response = httpx.post(
            _api_url(cfg.ingest_url, "/v1/policy/check"),
            headers=_headers(cfg),
            json={"intent_id": created_intent["id"]},
            timeout=_REQUEST_TIMEOUT_S,
        )
        policy_response.raise_for_status()
    except httpx.HTTPError as exc:
        raise ZrokyRuntimePolicyError(f"[ZROKY] Pre-execution guard unavailable: {exc}") from exc

    decision = policy_response.json()
    if decision.get("decision") == "allow":
        return {"intent": created_intent, "policy": decision}
    if decision.get("decision") == "approval_required":
        raise ZrokyRuntimePolicyApprovalRequired("[ZROKY] Pre-execution guard requires approval.", decision=decision)
    raise ZrokyRuntimePolicyBlocked("[ZROKY] Pre-execution guard did not allow action.", decision=decision)

File: zroky-sdk-js/src/intent.ts
// SPDX-License-Identifier: FSL-1.1-MIT
// Copyright 2026 Zroky AI

import { apiUrl, nodeEnv } from "./api";
import { resolveConfig } from "./config";
import {
  ZrokyRuntimePolicyApprovalRequired,
  ZrokyRuntimePolicyBlocked,
  ZrokyRuntimePolicyError,
  type RuntimePolicyDecision,
} from "./guard";
import type { ZrokyConfig } from "./types";

type JsonObject = Record<string, unknown>;

export interface PreExecutionGuardOptions {
  intent: JsonObject;
  environment?: string;
  agentRef?: string;
  idempotencyKey?: string;
}

export interface PreExecutionGuardResult {
  intent: JsonObject;
  policy: RuntimePolicyDecision;
}

function headers(apiKey: string, projectId: string, idempotencyKey?: string): Record<string, string> {
  return {
    "Content-Type": "application/json",
    "x-api-key": apiKey,
    "x-project-id": projectId,
    Authorization: `Bearer ${apiKey}`,
    ...(idempotencyKey ? { "Idempotency-Key": idempotencyKey } : {}),
  };
}

export async function preExecutionGuard(
  options: PreExecutionGuardOptions,
  config: ZrokyConfig = {},
): Promise<PreExecutionGuardResult> {
  const resolved = resolveConfig(config);
  const env = nodeEnv();
  const apiKey = resolved.apiKey ?? env?.["ZROKY_API_KEY"];
  const projectId = resolved.projectId ?? env?.["ZROKY_PROJECT_ID"] ?? env?.["ZROKY_PROJECT"];
  if (!apiKey || !projectId) {
    throw new ZrokyRuntimePolicyError("[ZROKY] Pre-execution guard requires apiKey and projectId.");
  }

  const key = options.idempotencyKey ?? crypto.randomUUID();
  const endpoint = resolved.endpoint ?? env?.["ZROKY_API_URL"] ?? env?.["ZROKY_ENDPOINT"];
  const intentResponse = await fetch(apiUrl(endpoint, "/v1/intents"), {
    method: "POST",
    headers: headers(apiKey, projectId, key),
    body: JSON.stringify({
      environment: options.environment ?? "production",
      agent_ref: options.agentRef,
      intent: options.intent,
    }),
  });
  if (!intentResponse.ok) {
    throw new ZrokyRuntimePolicyError(`[ZROKY] Trusted intent failed with HTTP ${intentResponse.status}.`);
  }
  const createdIntent = (await intentResponse.json()) as JsonObject;

  const policyResponse = await fetch(apiUrl(endpoint, "/v1/policy/check"), {
    method: "POST",
    headers: headers(apiKey, projectId),
    body: JSON.stringify({ intent_id: createdIntent.id }),
  });
  if (!policyResponse.ok) {
    throw new ZrokyRuntimePolicyError(`[ZROKY] Policy check failed with HTTP ${policyResponse.status}.`);
  }
  const policy = (await policyResponse.json()) as RuntimePolicyDecision;
  if (policy.decision === "allow") {
    return { intent: createdIntent, policy };
  }
  if (policy.decision === "approval_required") {
    throw new ZrokyRuntimePolicyApprovalRequired("[ZROKY] Pre-execution guard requires approval.", policy);
  }
  throw new ZrokyRuntimePolicyBlocked("[ZROKY] Pre-execution guard did not allow action.", policy);
}
